### ISSUE TITLE
Runtime Exception: Unable to write to process ID file. (Swoole + Docker)

### ISSUE THREAD

**Issue by Elycin** [laravel/octane#94]
- Octane Version: 0.1.0
- Laravel Version: 8.36.0
- PHP Version: 8.0.3
- Database Driver & Version: PostgreSQL (libpq) Version => 13.2

### Description:
Running a docker container, running laravel octane as root, shouldn't have permission issues.
```
  RuntimeException 

  Unable to write to process ID file.

  at vendor/laravel/octane/src/Swoole/ServerStateFile.php:62
     58▕      */
     59▕     public function writeState(array $newState): void
     60▕     {
     61▕         if (! is_writable($this->path) && ! is_writable(dirname($this->path))) {
  ➜  62▕             throw new RuntimeException('Unable to write to process ID file.');
     63▕         }
     64▕ 
     65▕         file_put_contents($this->path, json_encode(
     66▕             array_merge($this->read(), ['state' => $newState]),

      +27 vendor frames 
  28  artisan:37
      Illuminate\Foundation\Console\Kernel::handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
```

### Steps To Reproduce:
- Try to run octane in a docker container. (`ENTRYPOINT ['php', 'artisan', 'octane:start']`)

### Dockerfile to reproduce
```docker
FROM php:latest
RUN apt-get update && apt-get upgrade -y && apt-get install git -y
COPY ./ /app/
WORKDIR /app

# Install Extensions (Swoole)
ADD https://github.com/mlocati/docker-php-extension-installer/releases/latest/download/install-php-extensions /usr/local/bin/
RUN chmod +x /usr/local/bin/install-php-extensions && sync
RUN install-php-extensions swoole

# Install Source Code Dependencies
ADD https://getcomposer.org/composer-stable.phar /usr/local/bin/composer
RUN chmod +x /usr/local/bin/composer
RUN composer install

# Open Ports
EXPOSE 8000

# Entrypoint
ENTRYPOINT ["php", "artisan", "octane:start", "--host=0.0.0.0"]
```

**Comment by Elycin** [comment]
Out of curiosity, why not /tmp?

**Comment by Elycin** [comment]
There are a few potential reasons why I think this is problematic, maybe.
- SELinux/Apparmour
- is_writable only checks the user, not groups
- directory being provided may not exist

I've confirmed with a bash shell that the project directory is indeed writable, perhaps we can probably change this checking statement to validate the directory permission using chmod/masks?

**Comment by taylorotwell** [comment]
I can only provide feedback on using within Laravel Sail (which works correctly). Can't help debug custom Docker setups.

**Comment by Elycin** [comment]
Well, in that case:

For anyone else who is trying to get Laravel Octane into a Kubernetes cluster to make their project scale and may encounter this issue in the future: The problem seems rooted in the first execution. Simply popping into a shell and running octane:start twice caused the process to start running. **Find some way to run your command twice to get it working successfully**.

I hope maintainers are more optimistic rather than dismissive in the future when people report issues, because rather than what you might think at first glance it supposedly being a docker issue, the problem might actually be something else involving an edge case you might not consider, like a botched `sys_get_tmp_dir` call in PHP.

Video demonstration: https://www.youtube.com/watch?v=MwbfxS2i1Xo. 
(You will also see me create an application key as I cannot run octane without one, when I am initially providing it as an environment variable)

Insert obligatory "it works on my machine" sticker /s

**Comment by AlexGodbehere** [comment]
I'm encountering the same issue but the second restart doesn't help.

I'm trying to run Octane on a k8s cluster and receiving the `Unable to write to process ID file. at /app/vendor/laravel/octane/src/Swoole/ServerStateFile.php` error message as above. The aforementioned directory has the correct permissions and simply removing the checks in `ServerStateFile.php` sees Octane start up without fault.

I need to do some more digging on this but it looks like @Elycin's theory of `is_writable` only checking the user and not groups could be a good place to start.

**Comment by Elycin** [comment]
@AlexGodbehere I'm still working with my cluster, I'll be sure to keep you in the loop if I find something that works.

**Comment by Elycin** [comment]
@AlexGodbehere testing all morning on my M1 MacBook using Minikube.
The command I got to work still follows the principles of the above, but a little more merged/repititve.  
It did spit out "No arguments expected for `octane:start`, got `/dev/null`." however.

```
CMD php artisan octane:start 2>&1 /dev/null || php artisan octane:start --host=0.0.0.0 --port=8000
```

I'm still convinced the proper action for this issue should be to test that the library can actually write to `$this->path` rather than trusting `is_writable`, but I don't have much faith that this issue will be fixed for quite some time until someone opens a new issue with more information referencing this issue.

**Comment by AlexGodbehere** [comment]
@Elycin thanks for keeping me updated on this. I haven't had time to tinker myself but I'll update in here if I do.

**Comment by Elycin** [comment]
I wrote a proper docker image here testing using pure PHP, and I'm happy to say the problem isn't octane or docker itself, but rather the nature of PHP's `is_writable` method.

Using a junk PHP script I wrote out the following code and ran it into a docker container:
```php
<?php
// Example path of a file that we'll see if we can write to.
$file = "/var/tmp/file.txt";
$file_writable = is_writable($file);

// Get the directory of the path and test if the directory is writable.
$dir = dirname($file);
$dir_writable = is_writable($dir);

// Print result to console.
var_dump([
    "file_path" => $file,
    "file_path_is_writable" => $file_writable,
    "file_dir_path" => $dir,
    "file_dir_path_is_writable" => $dir_writable,
]);
```

This is the result I got. 
```php
array(4) {
  ["file_path"]=>
  string(17) "/var/tmp/file.txt"
  ["file_path_is_writable"]=>
  bool(false)
  ["file_dir_path"]=>
  string(8) "/var/tmp"
  ["file_dir_path_is_writable"]=>
  bool(true)
}
```

it's checking if the path of `$this->path` is writable, and what's happening here specifically is $this->path doesn't exist yet.
Digging deeper into Octane...

**Comment by Elycin** [comment]
Considering the thread dead at this point, maybe it has been fixed elsewhere, but I wanted to confirm for anyone else who may stumble across this that the solution still works.

```
NAME                                 READY   STATUS      RESTARTS   AGE
backend-6dc5f7557f-zq8st             1/1     Running     0          3h47m
backend-scheduler-1618815480-rk2pw   0/1     Completed   0          2m54s
backend-scheduler-1618815540-8fwwn   0/1     Completed   0          2m3s
backend-scheduler-1618815600-ctxsz   0/1     Completed   0          63s
backend-scheduler-1618815660-trqcv   0/1     Completed   0          3s
queueworker-7lg2h                    1/1     Running     0          4h49m
queueworker-hk8pb                    1/1     Running     0          3h28m
queueworker-t4xx8                    1/1     Running     0          4h49m
queueworker-vjnw2                    1/1     Running     0          3h28m
queueworker-x7cfs                    1/1     Running     0          4h49m
```

**Comment by AlexGodbehere** [comment]
@Elycin thanks for the update. I'm going to start looking into this more now that the NewRelic PHP agent for PHP8 is being released today.

**Comment by AlexGodbehere** [comment]
It looks as though the issue is indeed with the `is_writable` function as also experienced [here](https://github.com/nextcloud/server/issues/7124). Maybe there's a better way to check this?

**Comment by Elycin** [comment]
@AlexGodbehere Good luck with your findings! I had to move on to other things but hopefully, this will be fixed in the future. I see a lot of opportunity ahead.

**Comment by AlexGodbehere** [comment]
I managed to solve this. For me, it was a missing `/app/storage/logs` location. After creating the location in my Dockerfile with `RUN mkdir -p /app/storage/logs` Octane starts up first time.

**Comment by Elycin** [comment]
I'll be sure to add this to my dockerfile!

**Comment by Salman98** [comment]
I am using docker file in my case the container is running and showing http://127.0.0.1:8080 but when i try to access this , it don't work 


Docker command i am using to run container 
```
docker -d -p 8080:80 name
```

Docker file 
````
FROM php:latest

# Install dependencies and clear cache after.
# Note: we do all the installation in one command, because mostly the packages dose not change that freequently
# so the layer don't get rebuild and pulled often
RUN apt-get update \
 && apt-get install -y --no-install-recommends --fix-missing \
    git \
    zip \
    unzip \
    mariadb-client \
    libzip-dev \
    zlib1g-dev \
    libpng-dev \
    supervisor \
    libjpeg-dev \
    libpng-dev \
    locales \
    exiftool \
    libonig-dev \
    ffmpeg \
    cron \
    nano \
 && docker-php-ext-configure \
    gd --with-jpeg \
 && docker-php-ext-install \
    -j$(nproc) \
    pdo \
    pdo_mysql \
    mysqli \
    gd \
    zip \
    pcntl \
    exif \
    mbstring \
    pdo_mysql \
    opcache \
 && curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer \
 && rm -rf /var/lib/apt/lists/*
 #&& localedef -i en_US -c -f UTF-8 -A /usr/share/locale/locale.alias en_US.UTF-8

RUN pecl install swoole

RUN docker-php-ext-enable swoole

# Install dependencies and clear cache after.
# Note: we do all the installation in one comand, because mostly the packages dose not change that freequently
# so the layer dsn't get rebuild and pulled often

WORKDIR /var/www/html

CMD php artisan octane:install --server=swoole

# Configure supervisor
# Note: supervisor is running under root as you can see user=root, the superviser is used to start the laravel worker
# it runs 2 processors as you can see numprocs=2, this can be changed if the processing needs to be speed up
RUN \
# Create cron
echo "* * * * * cd /app && php artisan schedule:run >> /dev/null 2>&1" > /etc/cron.d/laravel-cron && \
chmod 0644 /etc/cron.d/laravel-cron && \
crontab /etc/cron.d/laravel-cron && \
\
if [ "$supervisor" = "true" ]; then \
echo "\
[program:laravel-worker] \n\
process_name=%(program_name)s_%(process_num)02d \n\
command=php /app/artisan queue:work --sleep=3 --tries=3 \n\
autostart=true \n\
autorestart=true \n\
user=root \n\
numprocs=2 \n\
redirect_stderr=true \n\
stdout_logfile=/app/storage/worker.log \n\
stopwaitsecs=3600 \n\
" > /etc/supervisor/conf.d/laravel-worker.conf \
&& touch /var/run/supervisor.sock \
&& chmod 777 /var/run/supervisor.sock \
&& echo "#/bin/bash\n\
/usr/bin/supervisord\
" > /usr/bin/app_services.sh; \
\
else \
\
# When supervisor is turned off we will create a empty ssh file
echo "#/bin/bash\n\
" > /usr/bin/app_services.sh; \
fi

RUN \
echo "\
file_uploads = On\n\
memory_limit = 2048M\n\
upload_max_filesize = 2048M\n\
upload_max_size = 2048M\n\
post_max_size = 2048M\n\
max_execution_time = 60\n\
" > /usr/local/etc/php/conf.d/uploads.ini

#COPY ./src/auth.json ./root/.composer/

# Copy source code
COPY ./src /var/www/html

## Set memory_limit
#ADD ./src/php/php.ini /usr/local/etc/php

## make sure enable in deployment
#COPY ./src/opcache.ini /usr/local/etc/php/conf.d/opcache.ini

RUN composer install ;
#--no-progress --no-scripts --optimize-autoloader --prefer-dist --no-dev

RUN chmod -R 777 ./storage && mkdir -p /app/storage/logs && php artisan key:generate;

#php  php artisan migrate && php artisan db:seed && php artisan key:generate

CMD  php artisan octane:start

````

**Comment by Radiergummi** [comment]
@AlexGodbehere that's indeed the solution here - there's simply _no reason at all_ to require the existence of a logs directory. Docker applications log to stdout/stderr as per best practices. This is nothing but a bug based on a naive assumption from the Laravel team.

**Comment by andrislusis** [comment]
I tried  to follow https://laravel.com/docs/8.x/octane#introduction documentation of enabling octane after installing fresh Laravel project with sail, did not do any customization. The default setup on Linux gives this error. Also other permission errors regarding storage and logs directory are there out of the box.

**Comment by jmvcollaborator** [comment]
Guys after digging further in the vendor files, I did check the value of 
$this->path in my case /var/www/{MYLARAVELPROJECT}/storage/logs 
and i realized that the directory logs was not writable for "others" drwxrwxr**-**x
just change others to have w (write permissions) drwxrwxr**w**x

Issue solved !!!

**Comment by shifas05** [comment]
`'state_file' => storage_path('logs/octane-server-state.json')` 
add above line in config/octane.php , it will work without any issues

**Comment by erycson-cubo** [comment]
> `'state_file' => storage_path('logs/octane-server-state.json')` adicione a linha acima em config/octane.php , está funcionando para mim

This worked for me, the error is in the OctaneServiceProvider.php file, where, when calling storage_path(), the path /tmp/storage/logs/octane-server-state.json is returned. And as PHP checks if it exists, before knowing if it is writable, the error ends up happening since the folder does not really exist.

File: OctaneServiceProvider.php
```php
$this->app->bind(SwooleServerStateFile::class, function ($app) {
    return new SwooleServerStateFile($app['config']->get(
        'octane.state_file',
        storage_path('logs/octane-server-state.json')
    )
    );
});
```

**Comment by erycson** [comment]
> `'state_file' => storage_path('logs/octane-server-state.json')` add above line in config/octane.php , its working for me

This worked for me, the error is in the OctaneServiceProvider.php file, where, when calling storage_path(), the path /tmp/storage/logs/octane-server-state.json is returned. And as PHP checks if it exists, before knowing if it is writable, the error ends up happening since the folder does not really exist.

File: OctaneServiceProvider.php
```php
$this->app->bind(SwooleServerStateFile::class, function ($app) {
    return new SwooleServerStateFile($app['config']->get(
        'octane.state_file',
        storage_path('logs/octane-server-state.json')
    )
    );
});
```

**Comment by ilyazastrognov** [comment]
Yes, those problems still exist. Laravel Octane is complete trash. That’s why I switched to Spiral, even though I really love Laravel.
Building a cluster with good scaling on Octane, while also integrating Laravel with plugins written in Go, is an impossible task for Octane. At least in our opinion.
A framework that runs on FPM is only suitable for brochure-style websites.

We spent so much time just trying to make Octane somewhat viable. In the end, we gave up and moved to a framework that works natively with RoadRunner.
